src/python/gem5/utils/looppoint.py

```diff
@@ -0,0 +1,404 @@
+    in workload
+
+    """
+
+    def __init__(
+        self,
+        targets: List[PcCountPair],
+        regions: Dict[PcCountPair, int],
+        json_file: Dict[int, Dict],
+    ) -> None:
+        """
+        :param targets: a list of PcCountPair that are used to generate exit
+        event at when the PcCountTrackerManager encounter this PcCountPair in
+        execution
+        :param regions: a dictionary used to find the corresponding region id
+        for the significant PcCountPair. This is mainly used to ensure
+        checkpoints are taken in the correct PcCountPair or relative counts are
+        updated at the correct count
+        :param json_file: all the LoopPoint data including relative counts and
+        multiplier are stored in this parameter. It can be outputted as a json
+        file.
+        """
+
+        self._manager = PcCountTrackerManager()
+        self._manager.targets = targets
+        self._targets = targets
+        self._regions = regions
+        self._json_file = json_file
+
+    def setup_processor(
+        self,
+        processor: AbstractProcessor,
+    ) -> None:
+        """
+        This function is used to setup a PC tracker in all the cores and
+        connect all the tracker to the PC tracker manager to perform
+        multithread PC tracking
+        :param processor: the processor used in the simulation configuration
+        """
+        for core in processor.get_cores():
+            core.add_pc_tracker_probe(self._targets, self._manager)
+
+    def update_relatives(self) -> None:
+        """
+        This function is used to update the relative count for restore used.
+        The new relative count will be stored in the _json_file and can be
+        outputted into a json file by calling the output_json_file function.
+        """
+        current_pair = self._manager.get_current_pc_count_pair()
+        temp_pair = PcCountPair(current_pair.getPC(), current_pair.getCount())
+        if temp_pair in self._regions:
+            rid = self._regions[temp_pair]
+            region = self._json_file[rid]["simulation"]
+            if "warmup" in self._json_file[rid]:
+                # if this region has a warmup interval,
+                # then update the relative count for the
+                # start of the simulation region
+                start = region["start"]["pc"]
+                temp = region["start"]["global"] - self._manager.get_pc_count(
+                    start
+                )
+                self._json_file[rid]["simulation"]["start"]["relative"] = int(
+                    temp
+                )
+            end = region["end"]["pc"]
+            temp = region["end"]["global"] - self._manager.get_pc_count(end)
+            self._json_file[rid]["simulation"]["end"]["relative"] = int(temp)
+
+    def output_json_file(
+        self, input_indent: int = 4, filename: str = "LoopPoint.json"
```

Comment:
Default should be machine, not human readable. I suggest making the default no indent.

For the filename, this should output a file relative to `outdir`.